third_party/move/move-compiler-v2/src/pipeline/variable_coalescing.rs

```diff
@@ -0,0 +1,239 @@
+//! local is `[0, MAX_CODE_OFFSET]`, but we can often compute more precise live intervals.
+//!
+//! The transformation greedily reuses (same-typed) locals outside their live intervals.
+
+use crate::pipeline::livevar_analysis_processor::LiveVarAnnotation;
+use move_binary_format::file_format::CodeOffset;
+use move_model::{ast::TempIndex, model::FunctionEnv, ty::Type};
+use move_stackless_bytecode::{
+    function_target::{FunctionData, FunctionTarget},
+    function_target_pipeline::{FunctionTargetProcessor, FunctionTargetsHolder},
+    stackless_bytecode::Bytecode,
+};
+use std::collections::{BTreeMap, BTreeSet};
+
+/// The live interval of a local (inclusive).
+/// Note that two live intervals i1: [b1, x] and i2: [x, e2] are *not* considered to overlap
+/// even though the code offset `x` is in both intervals.
+struct LiveInterval {
+    begin: CodeOffset,
+    end: CodeOffset,
+}
+
+impl LiveInterval {
+    /// Create a new live interval that only has the given offset.
+    fn new(offset: CodeOffset) -> Self {
+        Self {
+            begin: offset,
+            end: offset,
+        }
+    }
+
+    /// Include the given offset in the live interval, expanding the interval as necessary.
+    fn include(&mut self, offset: CodeOffset) {
+        self.begin = std::cmp::min(self.begin, offset);
+        self.end = std::cmp::max(self.end, offset);
+    }
+}
+
+/// Live interval event of a local, used for sorting.
+enum LiveIntervalEvent {
+    Begin(
+        /* which local? */ TempIndex,
+        /* live interval begins */ CodeOffset,
+        /* live interval length */ usize, // used for tie-breaking
+    ),
+    End(
+        /* which local? */ TempIndex,
+        /* live interval ends */ CodeOffset,
+    ),
+}
+
+impl LiveIntervalEvent {
+    /// Get the code offset at which the event occurs.
+    fn offset(&self) -> CodeOffset {
+        match self {
+            LiveIntervalEvent::Begin(_, offset, _) => *offset,
+            LiveIntervalEvent::End(_, offset) => *offset,
+        }
+    }
+}
+
+pub struct VariableCoalescing {}
+
+impl VariableCoalescing {
+    /// Compute the live intervals of locals in the given function target.
+    /// The result is a vector of live intervals, where the index of the vector is the local.
+    /// If a local has `None` as its live interval, we can ignore the local for the coalescing
+    /// transformation (eg., because it is borrowed or because it is never used): it implies
+    /// that it is the trivial live interval.
+    fn live_intervals(target: &FunctionTarget) -> Vec<Option<LiveInterval>> {
+        let LiveVarAnnotation(live_var_infos) = target
+            .get_annotations()
+            .get::<LiveVarAnnotation>()
+            .expect("live var annotation is a prerequisite");
+        // Note: we currently exclude all the variables that are borrowed from participating in this
+        // transformation, which is safe. However, we could be more precise in this regard.
```

Comment:
Yes, it really shouldn't be that difficult to mimic a better approximation of when a local is borrowed. This is not that complex has the full borrow analysis. One would kind of need to refcount how many references which are alive point to it. We should add this as a lightweight analysis such that it can be reused many times like livevar.

Reply:
Sounds reasonable. Will add in a followup PR - I would consider this a precision improving change and should not the fundamental framework setup for coalescing here.